naga/src/back/glsl/mod.rs

```diff
@@ -2077,42 +2083,110 @@ impl<'a, W: Write> Writer<'a, W> {
                 selector,
                 ref cases,
             } => {
-                // Start the switch
-                write!(self.out, "{level}")?;
-                write!(self.out, "switch(")?;
-                self.write_expr(selector, ctx)?;
-                writeln!(self.out, ") {{")?;
-
-                // Write all cases
                 let l2 = level.next();
-                for case in cases {
-                    match case.value {
-                        crate::SwitchValue::I32(value) => write!(self.out, "{l2}case {value}:")?,
-                        crate::SwitchValue::U32(value) => write!(self.out, "{l2}case {value}u:")?,
-                        crate::SwitchValue::Default => write!(self.out, "{l2}default:")?,
-                    }
+                // Some glsl consumers may not handle switches with a single body correctly.
+                // So write this as a `do {} while(false);` loop instead.
+                // This is the same fix as for FXC in hlsl, although here it
+                // wasn't specifically tested against any problematic consumers.
+                // See https://github.com/gfx-rs/wgpu/issues/4514
+                let one_body = cases
+                    .iter()
+                    .rev()
+                    .skip(1)
+                    .all(|case| case.fall_through && case.body.is_empty());
```

Comment:
One issue I just realized is what if none of these switch cases would have matched, I need to check if that's possible.

Reply:
Nvm looks like naga's switch statement requires a default case!

Reply:
Maybe I should note that here?

Reply:
I added a note